### PR TITLE
Preserve whitespace in text surrounded by newlines

### LINKED ISSUE #815: 🐛 BUG: HTML minification removes whitespace that should not be removed
### What version of `@astrojs/compiler` are you using?

1.5.3

### What package manager are you using?

pnpm

### What operating system are you using?

Linux (NixOS)

### Describe the Bug

Here's an example:

```html
<p>
    A
    <span>B</span>
    C
</p>
```

| Value of `compressHTML` | Screenshot of result |
|---|---|
|`false`|image|
|`true`|image|

The whitespace is not removed if "B" is also text. I have several more examples in the repo I provided.

### Link to Minimal Reproducible Example

https://gitlab.com/wackbyte/astro-compress-html-removes-whitespace

### PR DESCRIPTION
**Pull request by wackbyte**
## Changes

Closes #815. This issue came about as withastro/astro#7401 reported the previous whitespace-preserving behavior as incorrect, leading to #809. See [my comment](https://github.com/withastro/astro/issues/7401#issuecomment-1619327198) for why the behavior introduced by that PR is incorrect.

TLDR: revert #809 (& more)

## Testing

I added a new test based on #815 and fixed a few problems with old tests.

## Docs

Documentation not changed as this is purely a bugfix.

### CODE REVIEW

**Comment by wackbyte** [comment]
(It would also be possible to drop `hasNewline` and instead append/prepend only spaces)

**Review by natemoo-re — APPROVED** [review]
Thanks for the fix and for updating those tests!

Totally agree with [your comment](https://github.com/withastro/astro/issues/7401#issuecomment-1619327198) about the intended behavior. Sorry for the regression!
